File: app/src/main/java/dev/ragnarok/fenrir/adapter/feedback/FeedbackAdapter.java
package dev.ragnarok.fenrir.adapter.feedback;

import android.app.Activity;
import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.util.Calendar;
import java.util.EventListener;
import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.AttachmentsViewBinder;
import dev.ragnarok.fenrir.adapter.base.RecyclerBindableAdapter;
import dev.ragnarok.fenrir.model.feedback.CommentFeedback;
import dev.ragnarok.fenrir.model.feedback.CopyFeedback;
import dev.ragnarok.fenrir.model.feedback.Feedback;
import dev.ragnarok.fenrir.model.feedback.FeedbackType;
import dev.ragnarok.fenrir.model.feedback.LikeCommentFeedback;
import dev.ragnarok.fenrir.model.feedback.LikeFeedback;
import dev.ragnarok.fenrir.model.feedback.MentionCommentFeedback;
import dev.ragnarok.fenrir.model.feedback.MentionFeedback;
import dev.ragnarok.fenrir.model.feedback.PostPublishFeedback;
import dev.ragnarok.fenrir.model.feedback.ReplyCommentFeedback;
import dev.ragnarok.fenrir.model.feedback.UsersFeedback;
import dev.ragnarok.fenrir.view.OnlineView;

public class FeedbackAdapter extends RecyclerBindableAdapter<Feedback, FeedbackAdapter.FeedbackHolder> {

    private static final int TYPE_COMMENTS = 0;
    private static final int TYPE_USERS = 1;

    private static final int HEADER_DISABLE = 0;
    private static final int HEADER_TODAY = 1;
    private static final int HEADER_THIS_WEEK = 2;
    private static final int HEADER_OLD = 3;
    private static final int HEADER_YESTERDAY = 4;

    private final Context mContext;
    private final FeedbackViewBinder mFeedbackViewBinder;
    private ClickListener mClickListener;

    public FeedbackAdapter(Activity context, List<Feedback> items, AttachmentsViewBinder.OnAttachmentsActionCallback attachmentsActionCallback) {
        super(items);
        mContext = context;
        mFeedbackViewBinder = new FeedbackViewBinder(context, attachmentsActionCallback);
    }

    @Override
    protected void onBindItemViewHolder(FeedbackHolder viewHolder, int position, int type) {
        Feedback item = getItem(position);
        Feedback previous = position > 0 ? getItem(position - 1) : null;

        int headerStatus = getHeaderStatus(previous, item.getDate());

        switch (headerStatus) {
            case HEADER_DISABLE:
                viewHolder.headerRoot.setVisibility(View.GONE);
                break;
            case HEADER_OLD:
                viewHolder.headerRoot.setVisibility(View.VISIBLE);
                viewHolder.headerText.setText(mContext.getString(R.string.dialog_day_older));
                break;
            case HEADER_TODAY:
                viewHolder.headerRoot.setVisibility(View.VISIBLE);
                viewHolder.headerText.setText(mContext.getString(R.string.dialog_day_today));
                break;
            case HEADER_YESTERDAY:
                viewHolder.headerRoot.setVisibility(View.VISIBLE);
                viewHolder.headerText.setText(mContext.getString(R.string.dialog_day_yesterday));
                break;
            case HEADER_THIS_WEEK:
                viewHolder.headerRoot.setVisibility(View.VISIBLE);
                viewHolder.headerText.setText(mContext.getString(R.string.dialog_day_ten_days));
                break;
        }

        if (viewHolder instanceof CommentHolder) {
            configCommentHolder(item, (CommentHolder) viewHolder);
        }

        if (viewHolder instanceof UsersHolder) {
            configUserHolder(item, (UsersHolder) viewHolder);
        }

        viewHolder.contentRoot.setOnClickListener(v -> {
            if (mClickListener != null) {
                mClickListener.onNotificationClick(item);
            }
        });
    }

    public void setClickListener(ClickListener clickListener) {
        mClickListener = clickListener;
    }

    private void configCommentHolder(Feedback notification, CommentHolder holder) {
        switch (notification.getType()) {
            case FeedbackType.WALL_PUBLISH:
                mFeedbackViewBinder.configWallPublishFeedback((PostPublishFeedback) notification, holder);
                break;
            case FeedbackType.WALL:
                mFeedbackViewBinder.configWallFeedback((PostPublishFeedback) notification, holder);
                break;
            case FeedbackType.MENTION:
                mFeedbackViewBinder.configMentionFeedback((MentionFeedback) notification, holder);
                break;
            case FeedbackType.REPLY_COMMENT:
                mFeedbackViewBinder.configReplyCommentFeedback((ReplyCommentFeedback) notification, holder);
                break;
            case FeedbackType.REPLY_TOPIC:
                mFeedbackViewBinder.configReplyTopicFeedback((ReplyCommentFeedback) notification, holder);
                break;
            case FeedbackType.REPLY_COMMENT_PHOTO:
                mFeedbackViewBinder.configReplyCommentPhotoFeedback((ReplyCommentFeedback) notification, holder);
                break;
            case FeedbackType.REPLY_COMMENT_VIDEO:
                mFeedbackViewBinder.configReplyCommentVideoFeedback((ReplyCommentFeedback) notification, holder);
                break;
            case FeedbackType.MENTION_COMMENT_POST:
                mFeedbackViewBinder.configMentionCommentsFeedback((MentionCommentFeedback) notification, holder);
                break;
            case FeedbackType.MENTION_COMMENT_PHOTO:
                mFeedbackViewBinder.configMentionCommentsPhotoFeedback((MentionCommentFeedback) notification, holder);
                break;
            case FeedbackType.MENTION_COMMENT_VIDEO:
                mFeedbackViewBinder.configMentionCommentsVideoFeedback((MentionCommentFeedback) notification, holder);
                break;
            case FeedbackType.COMMENT_POST:
                mFeedbackViewBinder.configCommentPostFeedback((CommentFeedback) notification, holder);
                break;
            case FeedbackType.COMMENT_PHOTO:
                mFeedbackViewBinder.configCommentPhotoFeedback((CommentFeedback) notification, holder);
                break;
            case FeedbackType.COMMENT_VIDEO:
                mFeedbackViewBinder.configCommentVideoFeedback((CommentFeedback) notification, holder);
                break;
        }
    }

    private void configUserHolder(Feedback notification, UsersHolder holder) {
        switch (notification.getType()) {
            case FeedbackType.LIKE_POST:
                mFeedbackViewBinder.configLikePostFeedback((LikeFeedback) notification, holder);
                break;
            case FeedbackType.LIKE_PHOTO:
                mFeedbackViewBinder.configLikePhotoFeedback((LikeFeedback) notification, holder);
                break;
            case FeedbackType.LIKE_VIDEO:
                mFeedbackViewBinder.configLikeVideoFeedback((LikeFeedback) notification, holder);
                break;
            case FeedbackType.LIKE_COMMENT_POST:
                mFeedbackViewBinder.configLikeCommentFeedback((LikeCommentFeedback) notification, holder);
                break;
            case FeedbackType.LIKE_COMMENT_TOPIC:
                mFeedbackViewBinder.configLikeCommentTopicFeedback((LikeCommentFeedback) notification, holder);
                break;
            case FeedbackType.FOLLOW:
                mFeedbackViewBinder.configFollowFeedback((UsersFeedback) notification, holder);
                break;
            case FeedbackType.FRIEND_ACCEPTED:
                mFeedbackViewBinder.configFriendAcceptedFeedback((UsersFeedback) notification, holder);
                break;
            case FeedbackType.LIKE_COMMENT_PHOTO:
                mFeedbackViewBinder.configLikeCommentForPhotoFeedback((LikeCommentFeedback) notification, holder);
                break;
            case FeedbackType.LIKE_COMMENT_VIDEO:
                mFeedbackViewBinder.configLikeCommentVideoFeedback((LikeCommentFeedback) notification, holder);
                break;
            case FeedbackType.COPY_POST:
                mFeedbackViewBinder.configCopyPostFeedback((CopyFeedback) notification, holder);
                break;
            case FeedbackType.COPY_PHOTO:
                mFeedbackViewBinder.configCopyPhotoFeedback((CopyFeedback) notification, holder);
                break;
            case FeedbackType.COPY_VIDEO:
                mFeedbackViewBinder.configCopyVideoFeedback((CopyFeedback) notification, holder);
                break;
        }
    }

    private int getItemViewType(Feedback notification) {
        switch (notification.getType()) {
            case FeedbackType.WALL:
            case FeedbackType.WALL_PUBLISH:
            case FeedbackType.MENTION:
            case FeedbackType.COMMENT_POST:
            case FeedbackType.MENTION_COMMENT_POST:
            case FeedbackType.COMMENT_PHOTO:
            case FeedbackType.MENTION_COMMENT_VIDEO:
            case FeedbackType.MENTION_COMMENT_PHOTO:
            case FeedbackType.COMMENT_VIDEO:
            case FeedbackType.REPLY_COMMENT:
            case FeedbackType.REPLY_COMMENT_PHOTO:
            case FeedbackType.REPLY_COMMENT_VIDEO:
            case FeedbackType.REPLY_TOPIC:
                return TYPE_COMMENTS;
            case FeedbackType.FOLLOW:
            case FeedbackType.FRIEND_ACCEPTED:
            case FeedbackType.COPY_POST:
            case FeedbackType.COPY_PHOTO:
            case FeedbackType.COPY_VIDEO:
            case FeedbackType.LIKE_POST:
            case FeedbackType.LIKE_PHOTO:
            case FeedbackType.LIKE_VIDEO:
            case FeedbackType.LIKE_COMMENT_POST:
            case FeedbackType.LIKE_COMMENT_PHOTO:
            case FeedbackType.LIKE_COMMENT_VIDEO:
            case FeedbackType.LIKE_COMMENT_TOPIC:
                return TYPE_USERS;
        }

        throw new IllegalArgumentException("Invalid feedback type: " + notification.getType());
    }

    @Override
    protected int getItemType(int position) {
        Feedback notification = getItem(position - getHeadersCount());
        return getItemViewType(notification);
    }

    @Override
    protected FeedbackHolder viewHolder(View view, int type) {
        switch (type) {
            case TYPE_COMMENTS:
                return new CommentHolder(view);
            case TYPE_USERS:
                return new UsersHolder(view);
        }

        return null;
    }

    @Override
    protected int layoutId(int type) {
        switch (type) {
            case TYPE_COMMENTS:
                return R.layout.item_feedback_comment;
            case TYPE_USERS:
                return R.layout.item_feedback_user;
        }

        return 0;
    }

    private int getHeaderStatus(Feedback previous, long date) {
        Long previousDate = null;
        if (previous != null) {
            previousDate = previous.getDate() * 1000;
        }
        int stCurrent = getStatus(date * 1000);
        if (previousDate == null) {
            return stCurrent;
        } else {
            int stPrevious = getStatus(previousDate);
            if (stCurrent == stPrevious) {
                return HEADER_DISABLE;
            } else {
                return stCurrent;
            }
        }
    }

    private int getStatus(long time) {
        Calendar current = Calendar.getInstance();
        current.set(current.get(Calendar.YEAR), current.get(Calendar.MONTH), current.get(Calendar.DATE), 0, 0, 0);
        long today = current.getTimeInMillis();
        long yesterday = today - 24 * 60 * 60 * 1000;
        long week = today - (10 * 24 * 60 * 60 * 1000);
        if (time >= today) {
            return HEADER_TODAY;
        } else if (time >= yesterday) {
            return HEADER_YESTERDAY;
        } else if (time >= week) {
            return HEADER_THIS_WEEK;
        } else return HEADER_OLD;
    }

    public interface ClickListener extends EventListener {
        void onNotificationClick(Feedback notification);
    }

    static class FeedbackHolder extends RecyclerView.ViewHolder {

        final View headerRoot;
        final View contentRoot;
        final TextView headerText;

        FeedbackHolder(View itemView) {
            super(itemView);
            headerRoot = itemView.findViewById(R.id.header_root);
            contentRoot = itemView.findViewById(R.id.content_root);
            headerText = itemView.findViewById(R.id.item_feedback_header_title);
        }
    }

    static class UsersHolder extends FeedbackHolder {

        final ImageView uAvatar;
        final TextView uName;
        final TextView uInfo;
        final TextView uTime;
        final OnlineView uChangable;
        final ImageView ivAttachment;

        UsersHolder(View root) {
            super(root);
            uAvatar = root.findViewById(R.id.item_friend_avatar);
            uName = root.findViewById(R.id.item_friend_name);
            uInfo = root.findViewById(R.id.item_additional_info);
            uTime = root.findViewById(R.id.item_friend_time);
            uChangable = root.findViewById(R.id.item_circle_friend_changable);
            ivAttachment = root.findViewById(R.id.item_feedback_user_attachment);
        }
    }

    static class CommentHolder extends FeedbackHolder {

        final ImageView cOwnerAvatar;
        final OnlineView cChangable;
        final TextView cOwnerName;
        final TextView cOwnerText;
        final TextView cOwnerTime;
        final ImageView cReplyOwnerAvatar;
        final TextView cReplyName;
        final TextView cReplyText;
        final TextView cReplyTime;
        final ViewGroup cReplyContainer;
        final View vAttachmentsRoot;
        final View vReplyAttachmentsRoot;
        final ImageView ivRightAttachment;

        CommentHolder(View root) {
            super(root);
            cOwnerAvatar = root.findViewById(R.id.item_comment_owner_avatar);
            cChangable = root.findViewById(R.id.item_circle_changable);
            cOwnerName = root.findViewById(R.id.item_comment_owner_name);
            cOwnerText = root.findViewById(R.id.item_comment_text);
            cOwnerTime = root.findViewById(R.id.item_comment_time);
            cReplyOwnerAvatar = root.findViewById(R.id.item_comment_reply_owner_avatar);
            cReplyName = root.findViewById(R.id.item_comment_reply_owner_name);
            cReplyText = root.findViewById(R.id.item_comment_reply_text);
            cReplyTime = root.findViewById(R.id.item_comment_reply_time);
            cReplyContainer = root.findViewById(R.id.comment_reply_feedback);
            vAttachmentsRoot = root.findViewById(R.id.item_feedback_comment_attachments_root);
            vReplyAttachmentsRoot = root.findViewById(R.id.item_reply_comment_attachments_root);
            ivRightAttachment = root.findViewById(R.id.item_feedback_comment_attachment);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/feedback/FeedbackAdapter.kt
package dev.ragnarok.fenrir.fragment.feedback

import android.app.Activity
import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.fragment.base.AttachmentsViewBinder.OnAttachmentsActionCallback
import dev.ragnarok.fenrir.fragment.base.RecyclerBindableAdapter
import dev.ragnarok.fenrir.fragment.feedback.FeedbackAdapter.FeedbackHolder
import dev.ragnarok.fenrir.model.feedback.CommentFeedback
import dev.ragnarok.fenrir.model.feedback.CopyFeedback
import dev.ragnarok.fenrir.model.feedback.Feedback
import dev.ragnarok.fenrir.model.feedback.FeedbackType
import dev.ragnarok.fenrir.model.feedback.LikeCommentFeedback
import dev.ragnarok.fenrir.model.feedback.LikeFeedback
import dev.ragnarok.fenrir.model.feedback.MentionCommentFeedback
import dev.ragnarok.fenrir.model.feedback.MentionFeedback
import dev.ragnarok.fenrir.model.feedback.PostPublishFeedback
import dev.ragnarok.fenrir.model.feedback.ReplyCommentFeedback
import dev.ragnarok.fenrir.model.feedback.UsersFeedback
import dev.ragnarok.fenrir.view.OnlineView
import java.util.Calendar

class FeedbackAdapter(
    context: Activity,
    items: MutableList<Feedback>,
    attachmentsActionCallback: OnAttachmentsActionCallback
) : RecyclerBindableAdapter<Feedback, FeedbackHolder>(items) {
    private val mContext: Context
    private val mFeedbackViewBinder: FeedbackViewBinder
    private var mClickListener: ClickListener? = null
    override fun onBindItemViewHolder(
        viewHolder: FeedbackHolder,
        position: Int,
        type: Int
    ) {
        val item = getItem(position)
        val previous = if (position > 0) getItem(position - 1) else null
        when (getHeaderStatus(previous, item.date)) {
            HEADER_DISABLE -> viewHolder.headerRoot.visibility = View.GONE
            HEADER_OLD -> {
                viewHolder.headerRoot.visibility = View.VISIBLE
                viewHolder.headerText.text = mContext.getString(R.string.dialog_day_older)
            }

            HEADER_TODAY -> {
                viewHolder.headerRoot.visibility = View.VISIBLE
                viewHolder.headerText.text = mContext.getString(R.string.dialog_day_today)
            }

            HEADER_YESTERDAY -> {
                viewHolder.headerRoot.visibility = View.VISIBLE
                viewHolder.headerText.text = mContext.getString(R.string.dialog_day_yesterday)
            }

            HEADER_THIS_WEEK -> {
                viewHolder.headerRoot.visibility = View.VISIBLE
                viewHolder.headerText.text = mContext.getString(R.string.dialog_day_ten_days)
            }
        }
        if (viewHolder is CommentHolder) {
            configCommentHolder(item, viewHolder)
        }
        if (viewHolder is UsersHolder) {
            configUserHolder(item, viewHolder)
        }
        viewHolder.contentRoot.setOnClickListener {
            mClickListener?.onNotificationClick(item)
        }
    }

    fun setClickListener(clickListener: ClickListener?) {
        mClickListener = clickListener
    }

    private fun configCommentHolder(notification: Feedback, holder: CommentHolder) {
        when (notification.type) {
            FeedbackType.WALL_PUBLISH -> mFeedbackViewBinder.configWallPublishFeedback(
                notification as PostPublishFeedback,
                holder
            )

            FeedbackType.WALL -> mFeedbackViewBinder.configWallFeedback(
                notification as PostPublishFeedback,
                holder
            )

            FeedbackType.MENTION -> mFeedbackViewBinder.configMentionFeedback(
                notification as MentionFeedback,
                holder
            )

            FeedbackType.REPLY_COMMENT -> mFeedbackViewBinder.configReplyCommentFeedback(
                notification as ReplyCommentFeedback,
                holder
            )

            FeedbackType.REPLY_TOPIC -> mFeedbackViewBinder.configReplyTopicFeedback(
                notification as ReplyCommentFeedback,
                holder
            )

            FeedbackType.REPLY_COMMENT_PHOTO -> mFeedbackViewBinder.configReplyCommentPhotoFeedback(
                notification as ReplyCommentFeedback,
                holder
            )

            FeedbackType.REPLY_COMMENT_VIDEO -> mFeedbackViewBinder.configReplyCommentVideoFeedback(
                notification as ReplyCommentFeedback,
                holder
            )

            FeedbackType.MENTION_COMMENT_POST -> mFeedbackViewBinder.configMentionCommentsFeedback(
                notification as MentionCommentFeedback,
                holder
            )

            FeedbackType.MENTION_COMMENT_PHOTO -> mFeedbackViewBinder.configMentionCommentsPhotoFeedback(
                notification as MentionCommentFeedback,
                holder
            )

            FeedbackType.MENTION_COMMENT_VIDEO -> mFeedbackViewBinder.configMentionCommentsVideoFeedback(
                notification as MentionCommentFeedback,
                holder
            )

            FeedbackType.COMMENT_POST -> mFeedbackViewBinder.configCommentPostFeedback(
                notification as CommentFeedback,
                holder
            )

            FeedbackType.COMMENT_PHOTO -> mFeedbackViewBinder.configCommentPhotoFeedback(
                notification as CommentFeedback,
                holder
            )

            FeedbackType.COMMENT_VIDEO -> mFeedbackViewBinder.configCommentVideoFeedback(
                notification as CommentFeedback,
                holder
            )
        }
    }

    private fun configUserHolder(notification: Feedback, holder: UsersHolder) {
        when (notification.type) {
            FeedbackType.LIKE_POST -> mFeedbackViewBinder.configLikePostFeedback(
                notification as LikeFeedback,
                holder
            )

            FeedbackType.LIKE_PHOTO -> mFeedbackViewBinder.configLikePhotoFeedback(
                notification as LikeFeedback,
                holder
            )

            FeedbackType.LIKE_VIDEO -> mFeedbackViewBinder.configLikeVideoFeedback(
                notification as LikeFeedback,
                holder
            )

            FeedbackType.LIKE_COMMENT_POST -> mFeedbackViewBinder.configLikeCommentFeedback(
                notification as LikeCommentFeedback,
                holder
            )

            FeedbackType.LIKE_COMMENT_TOPIC -> mFeedbackViewBinder.configLikeCommentTopicFeedback(
                notification as LikeCommentFeedback,
                holder
            )

            FeedbackType.FOLLOW -> mFeedbackViewBinder.configFollowFeedback(
                notification as UsersFeedback,
                holder
            )

            FeedbackType.FRIEND_ACCEPTED -> mFeedbackViewBinder.configFriendAcceptedFeedback(
                notification as UsersFeedback,
                holder
            )

            FeedbackType.LIKE_COMMENT_PHOTO -> mFeedbackViewBinder.configLikeCommentForPhotoFeedback(
                notification as LikeCommentFeedback,
                holder
            )

            FeedbackType.LIKE_COMMENT_VIDEO -> mFeedbackViewBinder.configLikeCommentVideoFeedback(
                notification as LikeCommentFeedback,
                holder
            )

            FeedbackType.COPY_POST -> mFeedbackViewBinder.configCopyPostFeedback(
                notification as CopyFeedback,
                holder
            )

            FeedbackType.COPY_PHOTO -> mFeedbackViewBinder.configCopyPhotoFeedback(
                notification as CopyFeedback,
                holder
            )

            FeedbackType.COPY_VIDEO -> mFeedbackViewBinder.configCopyVideoFeedback(
                notification as CopyFeedback,
                holder
            )
        }
    }

    private fun getItemViewType(notification: Feedback): Int {
        when (notification.type) {
            FeedbackType.WALL, FeedbackType.WALL_PUBLISH, FeedbackType.MENTION, FeedbackType.COMMENT_POST, FeedbackType.MENTION_COMMENT_POST, FeedbackType.COMMENT_PHOTO, FeedbackType.MENTION_COMMENT_VIDEO, FeedbackType.MENTION_COMMENT_PHOTO, FeedbackType.COMMENT_VIDEO, FeedbackType.REPLY_COMMENT, FeedbackType.REPLY_COMMENT_PHOTO, FeedbackType.REPLY_COMMENT_VIDEO, FeedbackType.REPLY_TOPIC -> return TYPE_COMMENTS
            FeedbackType.FOLLOW, FeedbackType.FRIEND_ACCEPTED, FeedbackType.COPY_POST, FeedbackType.COPY_PHOTO, FeedbackType.COPY_VIDEO, FeedbackType.LIKE_POST, FeedbackType.LIKE_PHOTO, FeedbackType.LIKE_VIDEO, FeedbackType.LIKE_COMMENT_POST, FeedbackType.LIKE_COMMENT_PHOTO, FeedbackType.LIKE_COMMENT_VIDEO, FeedbackType.LIKE_COMMENT_TOPIC -> return TYPE_USERS
        }
        throw IllegalArgumentException("Invalid feedback type: " + notification.type)
    }

    override fun getItemType(position: Int): Int {
        val notification = getItem(position - headersCount)
        return getItemViewType(notification)
    }

    override fun viewHolder(view: View, type: Int): FeedbackHolder {
        when (type) {
            TYPE_COMMENTS -> return CommentHolder(view)
            TYPE_USERS -> return UsersHolder(view)
        }
        throw UnsupportedOperationException()
    }

    override fun layoutId(type: Int): Int {
        when (type) {
            TYPE_COMMENTS -> return R.layout.item_feedback_comment
            TYPE_USERS -> return R.layout.item_feedback_user
        }
        throw UnsupportedOperationException()
    }

    private fun getHeaderStatus(previous: Feedback?, date: Long): Int {
        var previousDate: Long? = null
        if (previous != null) {
            previousDate = previous.date * 1000
        }
        val stCurrent = getStatus(date * 1000)
        return if (previousDate == null) {
            stCurrent
        } else {
            val stPrevious = getStatus(previousDate)
            if (stCurrent == stPrevious) {
                HEADER_DISABLE
            } else {
                stCurrent
            }
        }
    }

    private fun getStatus(time: Long): Int {
        val current = Calendar.getInstance()
        current[current[Calendar.YEAR], current[Calendar.MONTH], current[Calendar.DATE], 0, 0] = 0
        val today = current.timeInMillis
        val yesterday = today - 24 * 60 * 60 * 1000
        val week = today - 10 * 24 * 60 * 60 * 1000
        return when {
            time >= today -> {
                HEADER_TODAY
            }

            time >= yesterday -> {
                HEADER_YESTERDAY
            }

            time >= week -> {
                HEADER_THIS_WEEK
            }

            else -> HEADER_OLD
        }
    }

    interface ClickListener {
        fun onNotificationClick(notification: Feedback)
    }

    open class FeedbackHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val headerRoot: View = itemView.findViewById(R.id.header_root)
        val contentRoot: View = itemView.findViewById(R.id.content_root)
        val headerText: TextView = itemView.findViewById(R.id.item_feedback_header_title)

    }

    class UsersHolder(root: View) : FeedbackHolder(root) {
        val uAvatar: ImageView = root.findViewById(R.id.item_friend_avatar)
        val uName: TextView = root.findViewById(R.id.item_friend_name)
        val uInfo: TextView = root.findViewById(R.id.item_additional_info)
        val uTime: TextView = root.findViewById(R.id.item_friend_time)
        val uChangable: OnlineView = root.findViewById(R.id.item_circle_friend_changable)
        val ivAttachment: ImageView = root.findViewById(R.id.item_feedback_user_attachment)
    }

    class CommentHolder(root: View) : FeedbackHolder(root) {
        val cOwnerAvatar: ImageView = root.findViewById(R.id.item_comment_owner_avatar)
        val cChangable: OnlineView = root.findViewById(R.id.item_circle_changable)
        val cOwnerName: TextView = root.findViewById(R.id.item_comment_owner_name)
        val cOwnerText: TextView = root.findViewById(R.id.item_comment_text)
        val cOwnerTime: TextView = root.findViewById(R.id.item_comment_time)
        val cReplyOwnerAvatar: ImageView = root.findViewById(R.id.item_comment_reply_owner_avatar)
        val cReplyName: TextView = root.findViewById(R.id.item_comment_reply_owner_name)
        val cReplyText: TextView = root.findViewById(R.id.item_comment_reply_text)
        val cReplyTime: TextView = root.findViewById(R.id.item_comment_reply_time)
        val cReplyContainer: ViewGroup = root.findViewById(R.id.comment_reply_feedback)
        val vAttachmentsRoot: View = root.findViewById(R.id.item_feedback_comment_attachments_root)
        val vReplyAttachmentsRoot: View =
            root.findViewById(R.id.item_reply_comment_attachments_root)
        val ivRightAttachment: ImageView = root.findViewById(R.id.item_feedback_comment_attachment)

    }

    companion object {
        private const val TYPE_COMMENTS = 0
        private const val TYPE_USERS = 1
        private const val HEADER_DISABLE = 0
        private const val HEADER_TODAY = 1
        private const val HEADER_THIS_WEEK = 2
        private const val HEADER_OLD = 3
        private const val HEADER_YESTERDAY = 4
    }

    init {
        mContext = context
        mFeedbackViewBinder = FeedbackViewBinder(context, attachmentsActionCallback)
    }
}